File:  AVFoundation/avFoundationExample/Model/audioPlayer.m
//
//  audioPlayer.m
//  audioRecorder
//
//  Created by Luis Castillo on 10/26/16.
//  Copyright © 2016 LC. All rights reserved.
//

#import "audioPlayer.h"

@implementation audioPlayer

@synthesize audio_player, delegate;



-(void)playAudioFromData:(NSData *) data
{
    
    NSError * audioSessionError = nil;
    [[AVAudioSession sharedInstance] setCategory: AVAudioSessionCategoryPlayback error: &audioSessionError];
    if (audioSessionError != nil) {
        NSLog(@"%@", [audioSessionError localizedDescription]);
    }
    
    NSError * audioError;
    AVAudioPlayer * newPlayer = [[AVAudioPlayer alloc] initWithData:data error:&audioError];

    if (audioError != nil) {
        NSLog(@"%@", [audioError localizedDescription]);
    }
    
    self.audio_player = newPlayer;
    
    [audio_player prepareToPlay];
    [audio_player setDelegate:self];
    if (audio_player .prepareToPlay)
    {
        [self.audio_player play];
    }
    else
    {
        [delegate playerUnAbleToPlay];
    }
}//eom

-(void)playAudioFromURL:(NSURL *) url
{
    
    NSError * audioSessionError = nil;
    [[AVAudioSession sharedInstance] setCategory: AVAudioSessionCategoryPlayback error: &audioSessionError];
    if (audioSessionError != nil) {
        NSLog(@"%@", [audioSessionError localizedDescription]);
    }
    
    NSError * audioError;
    AVAudioPlayer * newPlayer = [[AVAudioPlayer alloc]initWithContentsOfURL:url error:&audioError];
    
    if (audioError != nil) {
        NSLog(@"%@", [audioError localizedDescription]);
    }
    
    self.audio_player = newPlayer;
    
    [audio_player prepareToPlay];
    [audio_player setDelegate:self];
    if (audio_player .prepareToPlay)
    {
        [self.audio_player play];
    }
    else
    {
        [delegate playerUnAbleToPlay];
    }
}//eom


#pragma mark player delegates
-(void)audioPlayerDecodeErrorDidOccur:(AVAudioPlayer *)player error:(NSError *)error
{
    NSLog(@"audioPlayerDecodeErrorDidOccur");
    NSLog(@"error: %@", error.localizedDescription);
    [delegate playerError:error];
    
    NSError * audioSessionError = nil;
    [[AVAudioSession sharedInstance] setActive: NO error: &audioSessionError];
    if (audioSessionError != nil) {
        NSLog(@"%@", audioSessionError.localizedDescription);
    }
}//eom

-(void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player successfully:(BOOL)flag
{
    NSLog(@"\n\naudioPlayerDidFinishPlaying");
    
    NSLog(@"%@", player.settings);
    NSLog(@"%@", player.format);
    
    [delegate playerFinish];
    
    NSError * audioSessionError = nil;
    [[AVAudioSession sharedInstance] setActive: NO error: &audioSessionError];
    if (audioSessionError != nil) {
        NSLog(@"%@", audioSessionError.localizedDescription);
    }
}//eom

@end

File: Audio/recordAndPlayAudio/recordAndPlayAudio/AudioPlayer.swift
//
//  AudioPlayer.swift
//  recordAndPlayAudio
//
//  Created by Luis Castillo on 5/10/16.
//  Copyright © 2016 LC. All rights reserved.
//

import UIKit
import AVFoundation

class AudioPlayer: NSObject, AVAudioPlayerDelegate
{
    private var _audioPlayer:AVAudioPlayer?
    
    private var _constants = AudioConstants .instance
    
    //MARK: - Play Audio
    func playAudio()
    {
        do
        {
            try AVAudioSession.sharedInstance().setCategory(AVAudioSessionCategoryPlayback)
            
             let url:NSURL = _constants .audioRecordPathNSURL
            
            if NSFileManager.defaultManager().fileExistsAtPath(url.path!)
            {
                _audioPlayer = try AVAudioPlayer(contentsOfURL: url)
            
                _audioPlayer?.numberOfLoops     = 0
                _audioPlayer?.meteringEnabled   = true
                _audioPlayer?.delegate          = self
                
                _audioPlayer?.play()
            }
        }
        catch
        {
            print(error)
        }
    }//eom
    
    func playAudioFromUrl(url:NSURL)
    {
        do
        {
            try AVAudioSession.sharedInstance().setCategory(AVAudioSessionCategoryPlayback)
            
            if NSFileManager.defaultManager().fileExistsAtPath(url.path!)
            {
                _audioPlayer = try AVAudioPlayer(contentsOfURL: url)
                
                _audioPlayer?.numberOfLoops     = 0
                _audioPlayer?.meteringEnabled   = true
                _audioPlayer?.delegate          = self
                
                _audioPlayer?.play()
            }
        }
        catch
        {
            print(error)
        }
    }//eom
    
    
    
    func pausePlay()
    {
        _audioPlayer?.pause()
    }//eom
    
    
    func stopPlay()
    {
        _audioPlayer?.stop()
    }//eom
    
    
    //MARK: - Delegates
    func audioPlayerDecodeErrorDidOccur(player: AVAudioPlayer, error: NSError?) {
        print(error?.localizedDescription)
    }//eom
    
    func audioPlayerBeginInterruption(player: AVAudioPlayer) {
        
    }//eom
    
    func audioPlayerDidFinishPlaying(player: AVAudioPlayer, successfully flag: Bool) {
        
        _audioPlayer?.delegate = nil
        _audioPlayer = nil
        
        if flag
        {
            print("successfully played audio!")
        }
    }//eom
    
}//eoc
